// @zowe/cli/lib/zosfiles/ZosFiles.definition.js
"use strict";
/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/
const zosmf_for_zowe_sdk_1 = require("@zowe/zosmf-for-zowe-sdk");
const Create_definition_1 = require("./create/Create.definition");
const Delete_definition_1 = require("./delete/Delete.definition");
const Invoke_definition_1 = require("./invoke/Invoke.definition");
const Download_definition_1 = require("./download/Download.definition");
const List_definition_1 = require("./list/List.definition");
const Upload_definition_1 = require("./upload/Upload.definition");
const Mount_definition_1 = require("./mount/Mount.definition");
const Unmount_definition_1 = require("./unmount/Unmount.definition");
const HMigrate_definition_1 = require("./hMigrate/HMigrate.definition");
const HRecall_definition_1 = require("./hRecall/HRecall.definition");
const Copy_definition_1 = require("./copy/Copy.definition");
const Rename_definition_1 = require("./rename/Rename.definition");
const ZosFiles_options_1 = require("./ZosFiles.options");
/**
 * This object defines the top level command group for zosfiles. This is not
 * something that is intended to be used outside of this npm package.
 *
 * @private
 */
const definition = {
    name: "zos-files",
    aliases: ["files"],
    type: "group",
    summary: "Manage z/OS data sets",
    description: "Manage z/OS data sets, create data sets, and more",
    children: [
        Create_definition_1.CreateDefinition,
        Delete_definition_1.DeleteDefinition,
        Invoke_definition_1.InvokeDefinition,
        HMigrate_definition_1.HMigrateDefinition,
        HRecall_definition_1.HRecallDefinition,
        Download_definition_1.DownloadDefinition,
        List_definition_1.ListDefinition,
        Upload_definition_1.UploadDefinition,
        Mount_definition_1.MountDefinition,
        Unmount_definition_1.UnmountDefinition,
        Copy_definition_1.CopyDefinition,
        Rename_definition_1.RenameDefinition
    ],
    passOn: [
        {
            property: "options",
            value: [...zosmf_for_zowe_sdk_1.ZosmfSession.ZOSMF_CONNECTION_OPTIONS, ...ZosFiles_options_1.ZosFilesOptionDefinitions],
            merge: true,
            ignoreNodes: [
                { type: "group" }
            ]
        }
    ]
};
module.exports = definition;
